### PR TITLE
Remove Maven optional dependencies support

### PR DESCRIPTION
**Pull request by melix**
### Context

Whenever we used the "improved POM support" experimental flag, Gradle was
interpreting optional dependencies as constraints. It meant that if a
hard dependency was found in the graph, and that another module had an
optional dependency on it, then we would conflict resolve between the two
versions.

It appears that this behavior causes some troubles, for example mentioned
in #5434. In Maven, optional dependencies are really never used. They are
pure documentation.

We also want to provide a better alternative to optional dependencies,
which will include a better model of optional features. Support for optional
dependencies coming from a Maven POM may be included as part of this work.

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/upgrade_version_4.adoc [hunk 1]
```diff
@@ -129,6 +129,11 @@ For now, by-convention config files in subprojects will override any in the root
  * `getEnabledDirectoryReportDestinations()`, `getEnabledFileReportDestinations()` and `getEnabledReportNames()` have all been removed from `org.gradle.api.reporting.ReportContainer`.
  * link:{javadocPath}/org/gradle/StartParameter.html#getProjectProperties--[StartParameter.projectProperties] and link:{javadocPath}/org/gradle/StartParameter.html#getSystemPropertiesArgs--[StartParameter.systemPropertiesArgs] now return immutable maps.
 
+[[changes_5.0]]
+== Changes in 5.0
+
+* <<#rel5.0:pom_optional_dependencies,Maven optional dependencies are no longer interpreted as dependency constraints>>
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/upgrade_version_4.adoc]: I would rephrase this to explain it was a feature available behind the `IMPROVED_POM_SUPPORT` otherwise it looks scary. A bit like you did in the other section.

Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/upgrade_version_4.adoc]: Or it might be left for later, when we mention here that `IMPROVED_POM_SUPPORT` became the default and list the promoted features.

Comment (melix) [reply subprojects/docs/src/docs/userguide/upgrade_version_4.adoc]: Yeah I had the same thought. I just wanted to have a minimal update of the docs here, knowing that we're going to enable improved POM support by default at some point. In which case I would conflate the 2 sections into one, and just say that as part of this work we abandoned the promotion of optional dependencies to constraints.

Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/upgrade_version_4.adoc]: 👍 Works for me
